Taplytics Integration

jsalaber opened the pull request.

samdozor commented on build.gradle:
```diff
@@ -10,11 +11,21 @@ buildscript {
 dependencies {
-    //TODO: Option 1: Replace with your Maven Central Dependency (preferred)
-    //compile 'com.example:example-artifact:1.2.3'
+    compile 'com.taplytics.sdk:taplytics:+@aar'
```
can you use "implementation" instead? Also - this unfortunately cannot be a wildcard. mParticle customers need to have reproducible builds.

samdozor commented on build.gradle:
```diff
@@ -10,11 +11,21 @@ buildscript {
-    //compile fileTree(dir: 'libs', include: ['*.jar'])
+    compile fileTree(dir: 'libs', include: ['*.jar', '*.aar'])
+    //Volley
+    compile 'com.android.volley:volley:+'
```
volley and retrofit are in your SDKs POM so you should be able to remove this.

Also - it makes me nervous that in your POM, you're specifying a wildcard - seems pretty dangerous for customers?

samdozor commented on build.gradle:
```diff
@@ -10,11 +11,21 @@ buildscript {
+    compile 'com.squareup.retrofit2:retrofit:+'
+
+    //Only include this if you wish to enable push notifications:
+    compile("com.google.android.gms:play-services-gcm:9.+")
```
customers will add this to their gradle file, it shouldn't be here.

samdozor commented on src/main/java/com/mparticle/kits/TaplyticsKit.java:
```diff
@@ -0,0 +1,300 @@
+
+    private HashMap<String, Object> mergeOptions(Map<String, Object> tlOptions, Map<String, Object> configuration) {
+        HashMap<String, Object> merged = new HashMap<>(configuration);
+        for (Map.Entry<String, Object> entry : tlOptions.entrySet()) {
```
if I'm reading this right it will crash by default because `tlOptions` is null? also customers can null it out.

samdozor commented on src/main/java/com/mparticle/kits/TaplyticsKit.java:
```diff
@@ -0,0 +1,300 @@
+
+    private String getAPIKey(Map<String, String> settings) {
+        final String apiKey = getSettings().get(API_KEY);
+        if (TextUtils.isEmpty(apiKey)) {
```
Can you use KitUtils.isEmpty instead? Otherwise unit tests fill failure as TextUtils isn't available without an emulator.

samdozor commented on src/main/java/com/mparticle/kits/TaplyticsKit.java:
```diff
@@ -0,0 +1,300 @@
+    private String getAPIKey(Map<String, String> settings) {
+        final String apiKey = getSettings().get(API_KEY);
+        if (TextUtils.isEmpty(apiKey)) {
+            throw new IllegalArgumentException(API_KEY);
```
can you use a better message, perhaps something like "Failed to initialized the Taplytics SDK - no API key has been configured."

samdozor commented on src/main/java/com/mparticle/kits/TaplyticsKit.java:
```diff
@@ -0,0 +1,300 @@
+    public List<ReportingMessage> logout() { return null; }
+
+    @Override
+    public void removeUserAttribute(String attribute) { }
```
should you be calling setUserAttribute with a null value here?

samdozor commented on src/main/java/com/mparticle/kits/TaplyticsKit.java:
```diff
@@ -0,0 +1,300 @@
+
+    @Override
+    public List<ReportingMessage> logEvent(CommerceEvent event) {
+        TransactionAttributes transactionAttributes = event.getTransactionAttributes();
```
you should check if this is a Purchase event here vs something else like an add to cart or impression, etc.

See here for an advanced example: https://github.com/mparticle-integrations/mparticle-android-integration-appboy/blob/master/src/main/java/com/mparticle/kits/AppboyKit.java#L151

samdozor commented on src/main/java/com/mparticle/kits/TaplyticsKit.java:
```diff
@@ -0,0 +1,300 @@
+
+    @Override
+    public void onApplicationForeground() {
+        Taplytics.logEvent(EventAppForeground);
```
Is this a common event to log that all customers want?

jsalaber commented:
hey @samdozor I've addressed the PR comments.

As for 1., that's used in some of our legacy code that isn't necessary for our integration with mParticle.

samdozor commented:
@jsalaber I think @peterjenkins noticed that it was added to the Taplytics iOS SDK - has that been there for a while then?

jsalaber commented:
@samdozor yes it's been there awhile

samdozor commented:
Looking good! there are some minor changes to be made (the README for example), which we'll take care of.

This has been approved and moved to its own repo: https://github.com/mparticle-integrations/mparticle-android-integration-taplytics
